[`BC`] Update `get_(text|image|audio|video)_features` methods to return `BaseModelOutputWithPooling`

tomaarsen opened the pull request:
# What does this PR do?
* Add return_dict to get_text_features, get_image_features, get_audio_features, get_video_features methods to return 'BaseModelOutputWithPooling' by default

Fixes #42401

The architectures supporting `get_image_features` are all extremely different, with wildly different outputs for the get_image_features methods: 
* 2d outputs,
* 3d outputs,
* lists of 2d outputs (due to non-matching shapes),
* existing 'return_attentions' resulting in returning 2-tuple,
* existing 'return_dict' resulting in returning 3-tuples (???),
* high quality image embeddings,
* low quality image embeddings,
* deepstack image embeddings,
* etc. etc. etc.

This PR aims to normalize these, so that users and third party libraries can perform inference on individual modalities (e.g. via `get_image_features`) despite the full model requiring multiple modalities.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@zucchini-nlp @ArthurZucker @Cyrilvallez

- Tom Aarsen

tomaarsen committed:
Add return_dict to get_text_features methods to allow returning 'BaseModelOutputWithPooling'

Added to all architectures except blip-2, which has a much different structure here. It uses 'Blip2TextModelWithProjection' to get these embeddings/features, but this class isn't as simple to use

tomaarsen committed:
Add return_dict to get_image_features methods to allow returning 'BaseModelOutputWithPooling'

Well, the architectures supporting get_image_features are all extremely different, with wildly different outputs for the get_image_features methods. 2d outputs, 3d outputs, lists of 2d outputs (due to non-matching shapes), existing 'return_attentions' resulting in returning 2-tuple, existing 'return_dict' resulting in returning 3-tuples (???), high quality image embeddings, low quality image embeddings, deepstack image embeddings, etc. etc. etc.

And I only went through like 70-80% of all architectures with get_image_features before I gave up.

Standardisation of all of these sounds like a lost cause.

tomaarsen committed:
make fixup

zucchini-nlp commented on src/transformers/models/chameleon/modeling_chameleon.py:
```diff
@@ -853,18 +853,29 @@ def get_image_tokens(self, pixel_values: torch.FloatTensor):
+
+        if return_dict:
+            return BaseModelOutputWithPooling(
+                last_hidden_state=image_embeddings,
```
with chameleon it is a bit vague. The vision quantizer could return `hidden_states` before quantizing them which i believe is the last hidden state we want

zucchini-nlp commented on src/transformers/models/qwen3_omni_moe/modeling_qwen3_omni_moe.py:
```diff
@@ -1936,6 +1942,14 @@ def get_image_features(self, pixel_values: torch.FloatTensor, image_grid_thw: Op
+        if return_dict:
+            return BaseModelOutputWithPooling(
+                last_hidden_state=image_embeds,
+                # pooler_output=image_features,  # NOTE: @Tom no pooled embeddings here
```
same thing here, `image_embeds` are actually pooled embeddings and the last hidden state is not returned from `visual`

zucchini-nlp reviewed (commented):
We discussed this internally and decided to add `last_hidden_states` to all models as the last state from vision block. The pooled embeddings will stay of different shapes as is

For the last hidden state the shapes are already more standardized, with a few major options. The only special cases might be qwen-like models where each image encoding has different sequence length and thus the outputs are concatenated as `length*dim`

tomaarsen committed:
Ignore discrepancies for pooler_output, focus on last_hidden_state

tomaarsen committed:
Update get_image_features for the missing architectures

tomaarsen committed:
Update all get_audio_features

tomaarsen committed:
Update get_video_features, except instructblipvideo

Should be fine though, as that  'get_video_features' doesn't live on the AutoModel class, but the AutoModelForConditionalGeneration class

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

tomaarsen committed:
Run ruff formatting

tomaarsen committed:
Patch Glm4v VisionModel forward with BaseModelOutputWithPooling

tomaarsen committed:
Patch instructblip, although backwards incompatibility stands

tomaarsen committed:
Patch Kosmos2 and Ovis2

tomaarsen committed:
Reformat Ovis2

tomaarsen committed:
Avoid now-deprecated return_attentions

zucchini-nlp commented on src/transformers/models/instructblip/modeling_instructblip.py:
```diff
@@ -1216,10 +1241,18 @@ def get_image_features(
+                vision_outputs=vision_outputs,
+                qformer_outputs=query_outputs,
+            )
+
```
wow, this is super weird, I always disliked InstructBLIP-style outputs! We would need to break it in the future, taking a note to myself

zucchini-nlp commented on src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py:
```diff
@@ -1784,6 +1808,9 @@ def get_audio_features(
 
+        if return_dict:
+            return audio_outputs
+
```
I like this pattern in QwenOmni with passing `return_dict` to the encoder and just returning the outputs. It would be great to adapt models for this way, I will add more in a big comment at the end

zucchini-nlp reviewed (commented):
Thanks a lot for the changes, I see there are a few tricky models that do not fit neatly with `BaseModelOutput`

To wrap it up, to msake this work firstly we need to ensure that all vision encoders are capable of returning dict in the way that PreTrainedModels do, i.e. by checking `config,return_dict` and returning attentions, hidden states, pooled output etc. Then we can ask `get_image_features` to return the same dict which was output by an encoder (optionally pooled output is updated in VLMs). That will preserve all fields of the vision encoder output

I think the current state of the PR is already doing it with a few non-standard models. I left comments under those models so lmk if that makes sense

tomaarsen committed:
Remove NumFrames

tomaarsen commented on src/transformers/models/glm4v/modeling_glm4v.py:
```diff
@@ -813,8 +817,15 @@ def forward(self, hidden_states: torch.Tensor, grid_thw: torch.Tensor, **kwargs)
-        return hidden_states
+        merged_hidden_states = self.merger(hidden_states)
+
+        if return_dict:
```
Would that not be a bit unexpected? I don't know if we sometimes have `return_dict` as False by default.

tomaarsen committed:
Proposal to simplify get_..._features via TransformersKwargs & check_model_inputs

The changes in check_model_inputs aren't the clearest/prettiest, but they work well for now.

tomaarsen committed:
Revert check_model_inputs, adopt can_return_tuple, accept BC on get_..._features methods

This commit updates all get_text_features methods, even blip_2, which was previously not yet attempted

tomaarsen committed:
Fix typo: can_return_dict -> can_return_tuple

tomaarsen committed:
Adopt can_return_tuple for many get_image_features

A handful of outliers that aren't updated yet, e.g. if there's 2+ ModelOutput classes that are viable, or the vq-based ones

For context, the other modeling file classes haven't been updated with the new get_..._features format, nor have the tests

tomaarsen committed:
Update all get_audio_features, some edge cases handled (e.g. gemma3n)

tomaarsen committed:
Update most get_video_features,  some edge case remain, e.g. instructblipvideo

tomaarsen commented:
For context, these are the TODOs at this point:
* [ ] Unfinished architectures
    * [x] `fuyu` `get_image_features`: I don't think Fuyu has a real Vision Encoder beyond just a single Linear
    * [ ] `instructblip` `get_image_features`: The new format misses the `query_outputs`/`qformer_outputs`, should use a new `ModelOutput` subclass somehow.
    * [ ] `instructblipvideo` `get_video_features`: See above
    * [ ] `kosmos2` `get_image_features`: The new format misses the `projection_attentions`, should use a new `ModelOutput` subclass somehow.
    * [ ] `ovis2` `get_image_features`: The new format misses the `visual_indicator_features`, should use a new `ModelOutput` subclass somehow.
    * [ ] `deepseek_vl_hybrid` `get_image_features`: This method produces both `low_res_vision_encodings` and `high_res_vision_encodings`, should use a new `ModelOutput` subclass somehow to combine them.
    * [ ] `chameleon` `get_image_features`: Update the VQVAE class to output the hidden states before quantization.
    * [ ] `emu` `get_image_features`: Update the VQVAE class to output the hidden states before quantization.
* [ ] Updated all architectures to accept the new output format
* [ ] Add and/or update tests for the new output format
* [ ] Update docstrings
* [ ] Update type hints

---

- Tom Aarsen

tomaarsen committed:
Patch Fuyu, just return BaseModelOutputWithPooling without pooler

The Fuyu architecture doesn't have an image encoder:
> Architecturally, Fuyu is a vanilla decoder-only transformer - there is no image encoder.

tomaarsen committed:
Introduce ModelOutput subclass for Chameleon, patch get_image_features

tomaarsen committed:
Update modeling files with new output formats for get_..._features

tomaarsen committed:
Update fast_vlm modeling forward from modular llava to remove image_sizes

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

tomaarsen committed:
Update colqwen2 its self.vlm.model.visual call to expect BaseModelOutput

tomaarsen committed:
Replace prior return_dict with check_model_inputs on qwen2_5_vl its VisionTransformer

tomaarsen committed:
Use BaseModelOutputWithProjectionAttentions for Kosmos2 to allow returning the projection attentions

tomaarsen committed:
Update Emu akin to Chameleon

tomaarsen committed:
Update the blip architectures with a naive fix

A better solution might be to remove the qformer etc. calls from the get_image/video_features and run those separately in the forward methods.

tomaarsen committed:
Convert remaining modulars (emu3, janus), patch emu3

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

tomaarsen committed:
Patch blip test

tomaarsen committed:
Update deepseek_vl using a new BaseModelOutputWithHighResVisionEncodings

tomaarsen committed:
Remove 'copied' for blip_2, instructblip and kosmos2 as they required custom changes

tomaarsen committed:
Patch qwen3_vl and qwen3_vl_moe, where I used last_hidden_state instead of pooler_output

tomaarsen committed:
Run repo-consistency

zucchini-nlp reviewed (commented):
Nice work, overall structure looks good to me and much cleaner than prev designs! I think the major comment is about `_can_record_outputs` for all models which now have a `check_model_inputs` decorator. Otherwise we aren't adding attentions/hidden_states in the output dict

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

tomaarsen committed:
Use kwargs["output_hidden_states"] = True to hardcode output_hidden_states where needed

tomaarsen committed:
Update new GlmAsr get_audio_features on ForConditionalGeneration

tomaarsen committed:
Run make style

tomaarsen commented on src/transformers/models/glm4v/modeling_glm4v.py:
```diff
@@ -776,7 +776,11 @@ def rot_pos_emb(self, grid_thw):
         return rotary_pos_emb, pos_ids
 
-    def forward(self, hidden_states: torch.Tensor, grid_thw: torch.Tensor, **kwargs) -> torch.Tensor:
+    @check_model_inputs(tie_last_hidden_states=False)
```
The superclass `Glm4vPreTrainedModel` here implements `_can_record_outputs`, so that should already be fine right?

zucchini-nlp replied:
`Glm4vVisionModel` uses different layers when collecting attentions and hidden states, and we would need to overwrite it

tomaarsen commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -267,7 +269,8 @@ def __init__(self, config: GraniteSpeechEncoderConfig):
         self.num_layers = config.num_layers
 
-    def forward(self, hidden_states: torch.Tensor):
+    @check_model_inputs(tie_last_hidden_states=False)
```
This class subclasses `nn.Module`, how should I best support `_can_record_outputs` here? Updating what superclass is used?

zucchini-nlp replied:
oh no 😿 Audio models are still under work, makes sense. We can make it a `PreTrainedModel` and add it in auto classes mapping. Though it's not really necessary for this PR imo, ig we can simply remove `check_model_inputs` and let audio team to standardize the models at some point

tomaarsen replied:
Removed check_model_inputs in b39b6d1c9c073394dd368f2bf3c32ab89086cbdd. I can also use can_return_tuple, but it might be problematic if the user has `return_dict=False` in the config.

tomaarsen committed:
Try to add _can_record_outputs to florence2

tomaarsen committed:
Override JanusVisionModel.forward to avoid bad q-former copy from Blip2

tomaarsen committed:
Import missing BaseModelOutput

tomaarsen committed:
Pop deprecated 'return_attentions', setting 'return_dict' won't be useful iiuc

tomaarsen committed:
Reintroduce kwargs filtering in llava etc. for safety re. image_sizes

We also don't need to incorporate code cleanup etc. in this PR, we should keep it as minimal as possible and leave these kinds of lines intact.

tomaarsen commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -182,7 +183,6 @@ def get_image_features(
         if vision_feature_select_strategy not in ["default", "full"]:
             raise ValueError(f"Unexpected select feature strategy: {self.config.vision_feature_select_strategy}")
 
-        kwargs = {k: v for k, v in kwargs.items() if v is not None}
```
Reverted all of these kwargs changes in d348d935530d53e727256eb77df9085fec5d45df, I think I should avoid making these "code quality" changes in this PR as issues can slip through, changes will be harder to find in PRs like these, and to keep this PR as small as possible.

tomaarsen committed:
Use BaseModelOutputWithPooling superclass consistently for custom get_..._features outputs

tomaarsen commented on src/transformers/models/qwen3_omni_moe/modular_qwen3_omni_moe.py:
```diff
@@ -102,6 +102,28 @@
+@dataclass
+class BaseModelOutputWithDeepstackFeatures(ModelOutput):
+    """
+    Base class for model's outputs that also contains a pooling of the last hidden states.
```
Well spotted, this one should have been 
```
@dataclass
@auto_docstring
class BaseModelOutputWithDeepstackFeatures(BaseModelOutputWithPooling):
    """
    deepstack_features (`List[torch.FloatTensor]`, *optional*):
        List of hidden-states (feature maps) from deepstack layers.
    """

    deepstack_features: Optional[list[torch.FloatTensor]] = None
```
Which I did use in another file: `modular_qwen3_vl`.

I also found some issues with Sam, fixed in 71ea85a2cf8c9a317007cc04c408fda78bc829e6.

tomaarsen committed:
Update Blip-2 family and its BaseModelOutputWithVisionQformerOutputs

To use both a vision_outputs and qformer_outputs as keys in the BaseModelOutputWithPooling subclass, despite some duplication.

tomaarsen commented on src/transformers/models/instructblip/modeling_instructblip.py:
```diff
@@ -1315,13 +1329,18 @@ def forward(
 
-        language_model_inputs, vision_outputs, query_outputs = self.get_image_features(
+        # NOTE: @Tom this is not particularly clean, e.g. having to manually remove qformer outputs from vision outputs
+        image_features: BaseModelOutputWithQformerOutputs = self.get_image_features(
```
Done in 8c59e9513362e701c94ffe68edbe376d853f6d93

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

tomaarsen commented on src/transformers/models/florence2/modeling_florence2.py:
```diff
@@ -542,12 +543,16 @@ def __init__(self, config: Florence2VisionConfig):
         self.post_init()
 
-    def forward(self, hidden_states: torch.Tensor, **kwargs):
+    @check_model_inputs(tie_last_hidden_states=False)
```
I think then I'll leave florence2 as-is with just the `hidden_states` if that's okay. I can use the `OutputRecorder`, but there's no class in the Florence modeling files that outputs the `attn_output, attn_weights`.

tomaarsen committed:
Update glm4v _can_record_outputs

tomaarsen committed:
Remove check_model_inputs in granite_speech

I could also use can_return_tuple, but this might be problematic if `return_dict=False` in the config

tomaarsen committed:
Run make style

tomaarsen committed:
Add _can_record_outputs to Ovis2VisionModel

tomaarsen committed:
Update get_text_features/get_video_features from pe_video

tomaarsen committed:
Update missing case on sam3

tomaarsen committed:
Update get_text_features type hints to Union[tuple, BaseModelOutputWithPooling]

Blip-2 and Clvp are the only exceptions

tomaarsen committed:
Add _can_record_inputs to qwen2_5_omni and qwen2_5_vl

tomaarsen committed:
Update get_image_features and get_video_features on ernie4_5_vl_moe

Can we even use BaseModelOutputWithPooling for these? It's a MoE model

tomaarsen committed:
Update get_image_features type hints to Union[tuple, BaseModelOutputWithPooling]

With a handful of exceptions

tomaarsen committed:
Remove @auto_docstring from pe_video, it's seemingly not used on that arch

(or well documented)

tomaarsen committed:
Update get_video_features type hints to Union[tuple, BaseModelOutputWithPooling]

Only exceptions for BaseModelOutputWithDeepstackFeatures

tomaarsen committed:
Fix pe_video import issue

tomaarsen committed:
Update forward, test, and docstring for sam3

tomaarsen committed:
Update get_audio_features type hints to Union[tuple, BaseModelOutputWithPooling]

Also update BaseModelOutput to BaseModelOutputWithPooling in several places, leaving room for a potential pooled embedding to be computed by get_audio_features

tomaarsen committed:
Add simple test case for get_text_features

Fails on CLIP, MetaCLIP, Siglip, Siglip2 as they use 'self.text_model = text_model.text_model', bypassing the TextModel that has `check_model_inputs` cc @zucchini-nlp related to #42564

tomaarsen committed:
First attempt to get get_image_features under test, still 26 failures

tomaarsen committed:
Resolve several test failures, progress still slow and inconsistent

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

zucchini-nlp commented on src/transformers/models/sam3/modeling_sam3.py:
```diff
@@ -2147,13 +2147,15 @@ def __init__(self, config: Sam3Config):
         input_ids: torch.LongTensor,
         attention_mask: torch.Tensor | None = None,
+        return_dict: bool = False,
         **kwargs: Unpack[TransformersKwargs],
```
does it really work with `return_dict=False`? The decorator will force set it to `True` at runtime, afaik

zucchini-nlp reviewed (approved):
Already looks good to me. When the tests are ready, I think we can request review from Arthur and merge 🚀

tomaarsen committed:
Split up get_..._features tests more, should be simpler to disable/customize specific parts per arch

tomaarsen committed:
Fix emu3 tests, also track non-temporal ResNet in hidden_states

tomaarsen committed:
Patch chameleon, emu3, ernie4_5, janus

tomaarsen committed:
Skip output_attentions for FastVLM, timm doesn't accept it

But I'm not sure how to handle the output_hidden_states case

tomaarsen committed:
Patch groupvit, instructblip, ovis2

plus style

tomaarsen committed:
Patch paddleocr_vl, qwen2_5_omni, qwen2_5_vl, qwen2_vl, and skip test for perception_lm

perception_lm is still problematic with output_hidden_states, akin to fast_vlm

tomaarsen committed:
Patch qwen3_omni_moe, sam family, edgetam

P.s. edgetam had incorrect _can_record_outputs

Now, all issues that remain with get_image_features are due to 1) CLIP family issue and 2) unclarity with expected output_hidden_states for timm-based models

tomaarsen commented on src/transformers/models/edgetam/modeling_edgetam.py:
```diff
@@ -1232,8 +1228,12 @@ def get_image_features(
 
-        return feature_maps, feature_maps_position_embeddings, vision_outputs.hidden_states, vision_outputs.attentions
+        # NOTE: @Tom I'm not 100% sure that the feature_maps/feature_maps_position_embeddings match the
+        # fpn hidden states/position encoding order, still have to double-check
```
Tackled in 13875af6f1f5a3ff1b4fea072bf5680b28d6a3f0 where I took care of the Sam and EdgeTam family

tomaarsen committed:
Kill now unused BaseModelOutputWithFeatureMaps

tomaarsen committed:
Remove left-over return_dict from prior attempt

tomaarsen committed:
Allow for output_hidden_states in theory, but skip impossible tests

The tests are failing as edgetam doesn't output hidden_states. It used to, because of a broken TimmWrapper in _can_return_outputs.

tomaarsen committed:
Introduce tests for get_audio_features, fixed all architectures

tomaarsen committed:
Introduce tests for get_video_features, only ernie4_5_vl_moe is failing

It's failing as the split_sizes gets made too small, such that the video_embeds doesn't sum to the split_sizes anymore. I'm not sure how to best tackle it.

I also removed the get_video_features from PaddleOCR_vl, as I don't think it's meant to be used with video

tomaarsen committed:
Call post_init on GraniteSpeechCTCEncoder, which was given a PreTrainedModel subclass

tomaarsen committed:
Update llava_onevision test suite, only create video pixel_values in new method

Instead of in the common one, as that negatively affects other tests (as there's no video tokens in the inputs_ids then)

tomaarsen committed:
Create custom video input for ernie4_5_vl_moe

tomaarsen committed:
Skip CLIP family tests; they don't support output_hidden_states/output_attentions due to bug

tomaarsen committed:
Breaking: update Blip2Model.get_text_features to no longer output logits

tomaarsen committed:
Satisfy test_num_layers_is_small test for align

tomaarsen committed:
Test against last_hidden_state against batch_size and hidden_size

19 failures, mostly if architectures merge the first dimension with e.g. num_frames for videos, or swap dimensions from the norm with the hidden_state at index 1 in a 4d-tensor

I don't think it's reasonable to expect these to be 'fixed', they would require drastic changes in the architectures or somewhat arbitrary changes in the post-processing of the hidden states.

tomaarsen committed:
Skip last_hidden_state shape tests for unusual cases

E.g. when batch_size is merged with num_frames or num_patches, or hidden_size is in index -3 instead of index -1

tomaarsen committed:
Update docstrings via auto_docstring for all get_..._features methods

Also add to e.g. aria.md to ensure that get_..._features methods are documented

tomaarsen committed:
Ensure all auto_doc arguments are documented

tomaarsen committed:
Remove redundant docstrings

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

tomaarsen committed:
Also patch the new glm_image for get_image_features/output_hidden_states

tomaarsen committed:
Update modular files as per check_docstring rules ...

... to avoid modular/check_docstring conflicts. Modular would propargate changes from modular to modeling files, and then check_docstring would complain and update the modeling files only. This created an unstable state where one of the two scripts was unhappy. I resolved this by manually tracking down the check_docstring issues in the modular files.

tomaarsen committed:
Update glm-image dates via fix-repo

tomaarsen commented:
This PR should be about ready for a final review. To summarize, the overall goal is to provide a more generalized output of the `get_(text|image|audio|video)_features` methods, such that they return a `BaseModelOutputWithPooling` or a subclass. This always contains a `last_hidden_state`, and optionally also `pooler_output`, `attentions` (depending on `output_attentions`), `hidden_states` (depending on `output_hidden_states`), and sometimes some other options, depending on the architecture.

The idea is that this is a "full-fledged" way to inference the model with a single modality, something which is often not possible with multimodal models (as they tend to require all modalities to be passed).

Many architectures have also been updated via `_can_record_outputs` & `ModelOutput` such that the "lower level" classes (e.g. a ...VisionEncoder, ...TextEncoder, ...VQVAE) can also return hidden states and attentions. 

I've added 12 test cases (3 for each modality) on the common ModelTester that tests:
1. whether the `get_(text|image|audio|video)_features` methods outputs an object with `last_hidden_state`, `pooler_output`, `attentions` (if the architecture supports it), and `hidden_states`, and that `last_hidden_state.shape[0] == batch_size` and `last_hidden_state.shape[-1] == hidden_size`
2. whether the number of `attentions` outputs matches expectations
3. whether the number of `hidden_states` outputs matches expectations

There's a handful of models that don't abide by the "index 0 is batch_size" and "index -1 is hidden_size", those tests have been skipped. Ideally, we would also normalize the hidden_states fully, but I think that could be too invasive.

If it's convenient with the big reviews, I can make comments for each of the architectures that I made significant changes in to cover why I made those changes. You can also just ask me via a review comment or in Slack.

As of the last commit, all of the core features that I was aiming for have been implemented, the documentation updated, and the tests written.

- Tom Aarsen

vasqu commented on docs/source/en/model_doc/glm_image.md:
```diff
@@ -16,7 +16,7 @@ limitations under the License.
 
 -->
-*This model was released on 2026-01-10 and added to Hugging Face Transformers on 2026-01-10.*
+*This model was released on 2026-01-10 and added to Hugging Face Transformers on 2026-01-13.*
```
Ig this comes from `make fix-repo`?

vasqu commented on src/transformers/models/altclip/modeling_altclip.py:
```diff
@@ -1223,11 +1219,12 @@ def get_image_features(
+        vision_outputs.pooler_output = self.visual_projection(pooled_output)
 
-        return image_features
+        return vision_outputs
```
Just to double check have they (the functions) been used elsewhere? Since we no expect different outputs (the dataclass vs a single tensor) - similarly elsewhere

vasqu commented on src/transformers/models/aya_vision/modeling_aya_vision.py:
```diff
@@ -216,7 +204,8 @@ def get_image_features(
         kwargs = {k: v for k, v in kwargs.items() if v is not None}
         # this is not memory efficient at all (output_hidden_states=True) will save all the hidden states.
-        image_outputs = self.vision_tower(pixel_values, output_hidden_states=True, **kwargs)
+        kwargs["output_hidden_states"] = True
```
Same here with forcing the kwarg, mentioning it a last time here

vasqu commented on src/transformers/models/blip/modeling_blip.py:
```diff
@@ -581,28 +583,26 @@ def get_text_features(
             input_ids=input_ids,
             attention_mask=attention_mask,
             position_ids=position_ids,
+            **kwargs,
```
Do we need `return_dict=True` here, just to be safe

tomaarsen replied:
Done in ed5c1364805ac23624ba29731dc44314f5a9126e

vasqu commented on src/transformers/models/blip_2/modeling_blip_2.py:
```diff
@@ -1113,36 +1122,32 @@ def get_text_features(
                 input_ids=input_ids,
                 attention_mask=attention_mask,
-                return_dict=True,
+                **kwargs,
```
Would not remove the dict, we should make sure that we really have the correct type

vasqu commented on src/transformers/models/edgetam/modular_edgetam.py:
```diff
@@ -188,7 +183,7 @@ def _init_weights(self, module):
     config_class = EdgeTamVisionConfig
     main_input_name = "pixel_values"
-    _can_record_outputs = {"hidden_states": TimmWrapperModel, "attentions": TimmWrapperModel}
+    _can_record_outputs = {}
```
So timm handles it automatically now? Can we add a comment here

vasqu commented on src/transformers/models/emu3/modular_emu3.py:
```diff
@@ -747,12 +767,11 @@ def encode(self, pixel_values: torch.Tensor, image_sizes: torch.Tensor):
         # b t c h w -> b c t h w
-        hidden_states = hidden_states.permute(0, 2, 1, 3, 4)
-        hidden_states = self.quant_conv(hidden_states)
+        conv_hidden_states = self.quant_conv(hidden_states.permute(0, 2, 1, 3, 4))
```
Just a nit, but since there are shape comments above id like to keep it

vasqu commented on src/transformers/models/ernie4_5_vl_moe/modular_ernie4_5_vl_moe.py:
```diff
@@ -1381,6 +1400,14 @@ def __init__(self, config):
+
+    @auto_docstring(custom_args=ERNIE4_5_VL_MOE_IMAGE_COMMON_CUSTOM_ARGS)
+    def get_image_features(self, **super_kwargs):
+        return super().get_image_features(**super_kwargs)
```
Do we really need this? Should inherit it or are the docstrings really different (which I doubt)

vasqu commented on src/transformers/models/florence2/modular_florence2.py:
```diff
@@ -1344,6 +1345,7 @@ class Florence2VisionPreTrainedModel(PreTrainedModel):
     _supports_flex_attn = True
 
     _can_compile_fullgraph = True
+    _can_record_outputs = {"hidden_states": Florence2VisionBlock}
```
Looks like there are also attentions that have been ignored before? https://github.com/huggingface/transformers/blob/a41491fce14b476d44bdc76acaf5b8bd1dadba4a/src/transformers/models/florence2/modular_florence2.py#L1097

zucchini-nlp commented on docs/source/en/model_doc/blip-2.md:
```diff
@@ -97,6 +97,7 @@ If you're interested in submitting a resource to be included here, please feel f
 [[autodoc]] Blip2ForConditionalGeneration
     - forward
     - generate
+    - get_image_features
```
btw, blip-2 also has a text/qformer feature getter. If we included images, imo we need to include those two as well

zucchini-nlp reviewed (commented):
Huge PR, thanks for working on it! 

My main comment is about `tie_last_hidden_state=False` in the newly added `check_model_outputs`. IMO we don't want `hidden_states[-1] != last_hidden_state` unless it is for BC

tomaarsen commented on src/transformers/models/audioflamingo3/modeling_audioflamingo3.py:
```diff
@@ -317,7 +322,7 @@ def get_input_embeddings(self) -> nn.Module:
         self.conv1 = value
 
-    @can_return_tuple
+    @check_model_inputs(tie_last_hidden_states=False)
```
I'm down with using the default `tie_last_hidden_states=True` on cases where we didn't return hidden_states to begin with, as there's no regression there either way 👍

tomaarsen committed:
FloatTensor -> LongTensor for image_tokens

tomaarsen committed:
Add simple last_hidden_state description, fix output typing of Gemma3nAudioEncoder.forward

tomaarsen committed:
Add missing `-> tuple | BaseModel...` on check_model_inputs

Using ``check_model_inputs[^\n]*\n\s*def forward\([^\)]*\):``

tomaarsen committed:
Ensure forward typing with check_model_inputs is `-> tuple | BaseModel...`

Using ``check_model_inputs[^\n]*\n\s*def forward\([^\)]+\) -> (?!tuple | )``

tomaarsen committed:
Undo accidental rename of Ovis2VisionAttention

tomaarsen committed:
Fix incorrect type hints for blip family

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

Also update get_image_features etc. for lighton_ocr

tomaarsen committed:
Patch get_image_features for lighton_ocr

tomaarsen committed:
Explicitly use Ovis2VisionAttention in Ovis2VisionEncoderLayer in modular

tomaarsen committed:
Update use of get_image_features for lighton_ocr

Forgot to run tests to verify that it worked, oops

vasqu commented on src/transformers/models/pe_audio/modular_pe_audio.py:
These models don't need a separate `get_xxx` function? Or are we leaving them for now?

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -863,13 +864,19 @@ def __init__(
-        """Convert pixel image to image features via the encoder and projector."""
+    @can_return_tuple
+    @auto_docstring
+    def get_image_features(
```
This model is an interesting case because it's an encoder-decoder model that can have image(/video?) inputs. Here we clearly define the vision embeddings but the text embeddings could be treated the same as in base t5. 

So I'm really unsure if we really would want the text embedding to be the full model as in t5; the encoder portion only makes more sense to me intuitively.

tomaarsen committed:
Rerun python utils/add_dates.py

Not sure which script removed the date... :/

vasqu commented on tests/models/x_clip/test_modeling_x_clip.py:
```diff
@@ -575,6 +576,13 @@ def test_model_from_pretrained(self):
         model = XCLIPModel.from_pretrained(model_name)
         self.assertIsNotNone(model)
 
+    def _video_features_prepare_config_and_inputs(self):
```
If we add exceptions, add small comments

vasqu commented on tests/models/ernie4_5_vl_moe/test_modeling_ernie4_5_vl_moe.py:
```diff
@@ -266,6 +266,27 @@ def test_inputs_embeds_matches_input_ids(self):
     def test_multi_gpu_data_parallel_forward(self):
         pass
 
+    def _video_features_prepare_config_and_inputs(self):
```
Oh yea, it has different split behavior between image and video :grimacing:

vasqu reviewed (commented):
Huge work!!! Super important

Mind that I started reviewing yesterday and only wrapping up just now so might be some repetitions / same comments as @zucchini-nlp. My main points are
- Docs <-- we need them, especially to clearly define what we expect as the embedding
- Encoder-decoder models like t5 <-- unsure what to expect as embedding but personally I'm in favor of the encoder only, are other models like bart included here then?
- The tests can be refactored a lot, we have a lot of repetition between all the modality tester functions <-- not super important but definitely worth to check out if you have the time
- The overridden test parts often have some comments but sometimes not, IMO it's important to have it everywhere to make sure an outsider can understand why we did this
- Set the attn implementation with `set_attn_implementation` not on the config
- There is some inconsistency between using `return_dict` or not. I don't think it changes behavior just wanted to note it

Yea that's mostly it. Sorry for this huge review, a lot is duplicated or for viz (to mark it happened elsewhere)

tomaarsen committed:
Remove tie_last_hidden_states=False from check_model_inputs from ...

forward methods that previously did not return a BaseModelOutput

tomaarsen commented on src/transformers/models/audioflamingo3/modular_audioflamingo3.py:
```diff
@@ -53,7 +57,12 @@ class AudioFlamingo3Encoder(Qwen2AudioEncoder):
+        "attentions": AudioFlamingo3Attention,
+    }
+
+    @check_model_inputs(tie_last_hidden_states=False)
```
Done in 0ce7baccbad77bfd8ae26bfd554068a07163a3ec

vasqu commented on docs/source/en/model_doc/aimv2.md:
Can we also add another file just for general guidance on this - autodocstring is one thing but just a general introduction would be nice as well. Doesn't have to big but giving a gist and detail what embeddings are expected etc

tomaarsen committed:
Revert accidental metaclip import change

tomaarsen commented on src/transformers/models/edgetam_video/modeling_edgetam_video.py:
```diff
@@ -116,16 +116,10 @@ def forward(self, hidden_states):
+class EdgeTamVideoVisionEncoderOutput(BaseModelOutputWithPooling):
     r"""
     last_hidden_state (`torch.FloatTensor` of shape `(batch_size, height, width, hidden_size)`):
         Sequence of hidden-states at the output of the last layer of the model.
```
I want to go through these and see if I can remove as many as possible to defer to the `auto_docstring`. I believe I kept the `last_hidden_state` in a few spots as the auto_docstring uses `of shape '(batch_size, sequence_length, hidden_size)'`.

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

ArthurZucker commented on src/transformers/models/blip/modeling_blip.py:
```diff
@@ -621,15 +621,15 @@ def get_image_features(
-        image_features = self.visual_projection(pooled_output)
-
-        return image_features
+        return vision_outputs
```
re on standardizing, if this is a common way -> we need self._visual_proj_name and self._vision_model_name (this way we never have to clutter modeling code with these functions!)

ArthurZucker reviewed (commented):
Thanks a lot! 

1. This needs to be documented in the migration guide, with motivations, breaking changes, and what to expect.
2. Whenever possible we wand to standardize and un-clutter the modeling code. Have a look at the way we define the `EmbeddingAccessMixin`. Similarly this is a feature we want to "hide" in the sense that its not primordial to the model

tomaarsen committed:
Add missing return_dict=True in get_..._features methods

tomaarsen committed:
Add `output_hidden_states=True` in InternVL get_image_features

Only if needed

tomaarsen commented on src/transformers/models/internvl/modular_internvl.py:
```diff
@@ -513,10 +512,11 @@ def get_image_features(
+            vision_features = vision_outputs.last_hidden_state
         else:
-            vision_features = self.vision_model(pixel_values=pixel_values).hidden_states[vision_feature_layer]
+            vision_features = vision_outputs.hidden_states[vision_feature_layer]
```
Yes, done in 3f0c7545c9bd19c84a2bbd3f0161a603d7be10df

tomaarsen committed:
Add missing docstring for llava_next_video get_video_features

tomaarsen committed:
Quick clean-up in _video_features_prepare_config_and_inputs test helper

tomaarsen committed:
model.set_attn_implementation instead of config._attn_implementation

Note:  There's about ~10 other places that use config._attn_implementation in this test file alone

tomaarsen committed:
Add simple docstring to some helper methods re. inputs.

It's not extremely useful I think, as it has to be somewhat generic due to the large differences in the architectures

tomaarsen committed:
Explain why get_..._features test inputs are overridden

tomaarsen committed:
Undo incorrect return_dict=True change in deepseek_vl_hybrid

I added return_dict to get_low_res_image_features and get_high_res_image_features calls, but these methods already set return_dict automatically

tomaarsen committed:
Revert accidental metaclip import change

tomaarsen committed:
Adopt **vision_outputs in instructblip, but mess remains

vasqu commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -199,7 +188,7 @@ def get_image_features(
         image_features = self.multi_modal_projector(selected_image_feature)
 
-        if "image_sizes" in kwargs:
+        if kwargs.get("image_sizes") is not None:
```
Can we clarify this in a small comment at least? Seems a bit special enough to consider this

ArthurZucker commented on src/transformers/models/fast_vlm/modular_fast_vlm.py:
```diff
@@ -266,7 +351,47 @@ def forward(self, **super_kwargs):
+        vision_feature_select_strategy = (
+            vision_feature_select_strategy
+            if vision_feature_select_strategy is not None
+            else self.config.vision_feature_select_strategy
```
move this code to `check_model_inputs` please

ArthurZucker commented on src/transformers/models/janus/modeling_janus.py:
```diff
@@ -424,136 +424,14 @@ def forward(
     config: JanusVisionConfig
     _can_record_outputs = {
-        "hidden_states": JanusEncoderLayer,
-        "attentions": JanusAttention,
```
nice catch I think here!

ArthurZucker commented on src/transformers/models/llama4/modeling_llama4.py:
```diff
@@ -1198,30 +1204,25 @@ def set_decoder(self, decoder):
         """
         if vision_feature_select_strategy not in ["default", "full"]:
-            raise ValueError(f"Unexpected select feature strategy: {self.vision_feature_select_strategy}")
+            raise ValueError(f"Unexpected select feature strategy: {vision_feature_select_strategy}")
```
`vision_feature_select_strategy` and related checks should be moved to check_model_inputs as its config related

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

ArthurZucker reviewed (approved):
Okay, let's get this merge! 

This is a good addition in terms of output recording and giving more freedom to users. 
It's not perfect and we're gonna try to hide the `get_image_features` when it is generic, but for now it works. 

Thanks for the hard work 🤗

tomaarsen commented on src/transformers/models/aria/modeling_aria.py:
```diff
@@ -1069,16 +1069,19 @@ def set_input_embeddings(self, value):
-    ):
+        **kwargs: Unpack[TransformersKwargs],
+    ) -> tuple | BaseModelOutputWithPooling:
         return self.model.get_image_features(
```
As discussed via Slack, we'll go with the current "manual" implementation initially to place the breaking change pre-v5, and post-v5 we can generalize these as much as possible in a new Mixin.

tomaarsen committed:
Avoid kwargs["output_hidden_states"] = True in get_..._features methods

tomaarsen commented on src/transformers/models/aria/modeling_aria.py:
```diff
@@ -910,42 +915,36 @@ def get_input_embeddings(self):
             vision_feature_layer if vision_feature_layer is not None else self.config.vision_feature_layer
         )
         patch_attention_mask = self._create_patch_attention_mask(pixel_mask)
+        kwargs["output_hidden_states"] = True
```
Done in a1e67675a3dcf338bbe9649edbf44f8b27fba829

tomaarsen committed:
Update check_model_inputs to default vision args based on config

tomaarsen committed:
Unrelated but important: patch set_attn_implementation for Windows

idem with set_experts_implementation

tomaarsen committed:
Revert output_hidden_states changes on InternVL

On this architecture, it seems cleaner to go the `kwargs["output_hidden_states"] = True` route, as a simple `output_hidden_states=vision_feature_layer != -1` prevents setting the `output_hidden_states` to True if requested for downstream use.

tomaarsen committed:
Extend d9001cc (check_model_inputs); remove more vision_feature_layer defaulting

tomaarsen committed:
Patch unusual bug: llava_next_video used self.vision_feature_layer

Doesn't seem like this was being used elsewhere, so I can just update it to use the local variant like elsewhere

tomaarsen committed:
Add unused use_cache to TimmWrapperModel to patch FastVLM

FastVLM now forwards this argument due to the check_model_inputs, and TimmWrapper can't use it

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

tomaarsen commented on tests/test_modeling_common.py:
```diff
@@ -4566,6 +4567,654 @@ def test_can_load_from_already_mapped_keys(self):
+        raise ValueError("Cannot determine the expected number of layers for video features")
+
+    def _video_features_get_expected_num_hidden_states(self, model_tester=None):
+        return self._video_features_get_expected_num_attentions(model_tester) + 1
```
I agree, but with v5 planned very soon, let's not worry about how pretty the tests are for now. They're sufficiently extensive (although sadly there are some skipped architectures), that works for me.

tomaarsen committed:
Update check_config_attributes to allow for vision attributes

And rerun fix-repo

tomaarsen committed:
Add tests for config.return_dict=False

Also; siglip had "nested" check_model_inputs: the VisionModel and VisionTransformer (below it) both used `check_model_inputs`. This means that the VisionModel.forward eats the 'return_dict=True', and the lower VisionTransformer.forward its `check_model_inputs` uses the config.return_dict=False to turn the output to a tuple.

The siglip/clip/metaclip family is still broken due to the `text_model = text_model.text_model` bypassing the class with the `check_model_inputs`.

tomaarsen committed:
permute and quantize separately for the comment

tomaarsen committed:
Ditch shared custom_args for ernie4_5_vl_moe

tomaarsen committed:
Move Ernie4_5_VL_MoeVisionAttention next to VisionBlock

tomaarsen committed:
Add missing "attentions" from Florence2 _can_record_outputs

tomaarsen committed:
Clarify kwargs.get("image_sizes") in modeling_llava

tomaarsen committed:
Remove commented skip_test_image_features_output_shape in chameleon tests

tomaarsen committed:
Add a migration guide under 'Library-wide changes with lesser impact'

tomaarsen commented:
Migration guide is ready. I don't feel like we should aim for docs here, the output formats are in my opinion too halfbaked to warrant advertising this as a feature apart from the API Reference.

- Tom Aarsen

vasqu commented on src/transformers/modeling_utils.py:
```diff
@@ -1909,7 +1909,7 @@ def _can_set_attn_implementation(cls) -> bool:
         """
         class_file = sys.modules[cls.__module__].__file__
-        with open(class_file, "r") as f:
+        with open(class_file, "r", encoding="utf8") as f:
```
Windows failing?

vasqu replied:
Would be nicer yes tbh. A bit unrelated to this PR, let's keep a clean history

vasqu commented on src/transformers/models/llava_next_video/modeling_llava_next_video.py:
```diff
@@ -395,41 +396,32 @@ def pack_image_features(self, image_features, image_sizes, vision_feature_select
+            The tensors corresponding to the input images.
+        image_sizes (`torch.Tensor` of shape `(num_images, 2)`)
+            Actual image size of each images (H, W).
+        vision_feature_layer (`Union[int, list[int]]`, *optional*):
```
same

vasqu reviewed (approved):
My last few nits ~

I'm fine with keeping this in the migration guide only. It is a bit sad but we should establish some kind of standard, e.g. like you mentioned qwen3 vl, in the future. While older models will be harder to fix for now, we can keep it in mind to not further "diversify" if possible.

tomaarsen commented on tests/test_modeling_common.py:
```diff
@@ -4634,6 +4635,722 @@ def test_can_load_from_already_mapped_keys(self):
+            self.assertEqual(last_hidden_state.shape, expected_shape, "last_hidden_state shape mismatch")
+
+            config, inputs_dict = self._text_features_prepare_config_and_inputs()
+            config.return_dict = False
```
You mean once with True and once with False? We can, I just went with this for now as the output checks are much different.

tomaarsen committed:
Parameterize get_..._features tests  with return_dict (True, False, None)

tomaarsen committed:
Add comment re. TimmWrapper _can_record_outputs

tomaarsen committed:
Shrink Gemma3nAudioEncoderModelOutput with auto_docstring & superclass

tomaarsen committed:
Revert "Unrelated but important: patch set_attn_implementation for Windows"

This reverts commit 092321671197d878e48b1c89edd154f47bb43a30.

tomaarsen committed:
Merge branch 'main' into feat/normalize_get_features_methods

Mecoli1219 commented:
Hi @tomaarsen, I’m testing the Llama 4 implementation in transformers (v5) and encountered the following error during the forward pass of `Llama4ForConditionalGeneration`:
```
AttributeError: 'BaseModelOutput' object has no attribute 'pooler_output'
```

After digging into the source, I suspect this might be a regression from this PR. It appears that `Llama4ForConditionalGeneration` now expects `get_image_features` to return a [pooler_output](https://github.com/huggingface/transformers/blob/b2028e775a52bf57ac2b6bd71b49ce61fa3adde6/src/transformers/models/llama4/modeling_llama4.py#L1327), but the underlying `Llama4VisionModel` (or the `get_image_features` method itself) still returns a standard `BaseModelOutput`.

Does this align with the intended architectural changes for Llama 4, or should `get_image_features` be updated to handle the issue safely?
